### PR TITLE
check if hardware context limit reached

### PR DESCRIPTION
**Pull request by entonbiba**
return false if audiocontext limit has been reached and console outputs the following message "This browser supports up to 6 hardware contexts".

Might need to add check for `console.log` or default to window.alert to return the message.

#1005

here is the chrome test check,
https://chromium.googlesource.com/experimental/chromium/blink/+/master/LayoutTests/webaudio/audiocontext-max-contexts-expected.txt

### CODE REVIEW

**Comment by katspaugh** [comment]
I think the request in #1005 was that the error message should be more meaningful, not how the exception is delivered. So just catch the error and throw your own exception with an instructive message.

**Comment by thijstriemstra** [comment]
Agreed.

**Comment by entonbiba** [comment]
@katspaugh @thijstriemstra what do you think of that solution?

**Comment by entonbiba** [comment]
@katspaugh @thijstriemstra what do you think of that solution to throw custom error message if it reaches 5 audio contexts instead of catching the error on the `this.ac = new().catch();`?

File: src/webaudio.js [hunk 1]
```diff
@@ -12,9 +12,14 @@ WaveSurfer.WebAudio = {
 
     getAudioContext: function () {
         if (!this.ac) {
-            this.ac = new (
-                window.AudioContext || window.webkitAudioContext
-            );
+            // check if audiocontext limit has been reached
+            if (this.ac.length >= 5) {
```
Comment (katspaugh) [inline src/webaudio.js]: I don't think we should hardcode any specific number. It probably depends on the browser. Just catch the error the browser is throwing when instantiating AudioContext.

**Comment by katspaugh** [comment]
> what do you think of that solution to throw custom error message

I've asked @mspae why we allow this error to happen in the first place. If it's really justified, then the message should be something like "You've created too many Web Audio contexts. Pass a single instance in the ac parameter". Sounds like bad UX anyway.

**Comment by mspae** [comment]
@katspaugh I wanted to remove the dependence on a global variable (and I might have run into other problems with `react-wavesurfer`, but I'm not sure what that was exactly)

I agree that there is probably no way round a global variable AFAIK in this case to get the proper singleton behaviour that the audiocontext API should be used with. (= there should only ever be one audiocontext ideally)

I will research some more on different aspects that need to be taken into account here (usability, "hot-rendering" of wavesurfer in for example react, server side rendering (= dependence on window variable), performance, extendability etc.) and get back to you with an update this evening (also concerning the "state of the union" for the v2 release since we're nearing the release date I envisioned in https://github.com/katspaugh/wavesurfer.js/pull/926#issuecomment-274932307) – Sorry for the delay but I was busy the last few days and want to give this the proper attention it deserves.

**Comment by katspaugh** [comment]
Awesome, thanks @mspae!

**Comment by mspae** [comment]
So I think we should
* reintroduce the audiocontext singleton as a global variable
* document how / add a new method / add a parameter to `wavesurfer.destroy` – to destroy the singleton for those who want that
* until we figure out how to port wavesurfer to the server (nodejs) I would recommend users disable loading of the wavesurfer library on the server. I think when projects like [web-audio-api](https://www.npmjs.com/package/web-audio-api) become more mature we could try and do that.

The v2 update of mine will follow friday. I didn't have enough time today as it turns out. :(

**Comment by mspae** [comment]
So I've also put together an example to test memory profiling by using lots and lots of wavesurfer instances with different options: https://github.com/mspae/wavesurfer.js/tree/audiocontext-singleton

… I won't open a PR obviously since this is for development only. If you want to give it a try start the normal dev task and go to `http://localhost:8080/example/audiocontext/?level=0&elements=100`
* `level` parameter reflects the state that should automatically be invoked `(['create html elements', 'create instances', 'load audio', 'destroy', 'close audio context'])`
* `elements` determines the number of instances
* `interval` is the wait interval between the different levels
* wavesurfer options can also be set as url parameters

This might be adjusted for memory profiling for other aspects of the library/plugins.

**Comment by mspae** [comment]
And here is my general v2 update:

### Things that have been done:
* HTML init script
* Added (more) complete jsdoc comments for the main library and customised esdoc for our needs. I would propose we use this to replace the documentation on the website (and provide it there) because it is probably considerably easier to maintain than a seperate documentation.
* ES6 class syntax has been implemented

### One PR to go ...
There is one architectural PR I would like to merge before we release the first beta IMHO (I'm currently working on it) – The issues the PR addresses are all to do with the plugin API:
* The plugin definition format includes two (!) properties that are actually reserved words in the JS language :rofl: (`extends` and `static`) – although this is not a problem within the current spec, old browsers might have problems with that ([stackoverflow](http://stackoverflow.com/questions/7022397/using-reserved-words-as-property-names-revisited))
* It would be nice to use esdoc generated documentation in plugins, esdoc only supports documentation for classes which are actually exported, the current plugin factory approach doesn't work.
* I feel the plugin inheritance mechanism is the unstable-est part of the lib right now and want to fix that. – Also it's basically used to make things into and observer AND in the minimap plugin it is used to inherit from the `Drawer`

… So the upcoming PR will introduce the following changes:
* There is no `extends` property anymore in the plugin definition format
* All plugins automatically inherit from the observer
* Minimap instantiates a drawer using `this.drawer = wavesurfer.Drawer(el, params);` and doesn't directly inherit from drawer.
* Plugin files export the plugin definition factory and the plugin class. This has the `@inherits {Observer}` jsdoc tag so the documentation generator gets things right.
* Plugin class constructor functions are called with `wavesurfer` and `params` parameters by the core lib

I think this is a good middle ground between a simple API which provides for 99% of the use cases and still allows pretty advanced features.

## 'Nice to haves' which can happen after first beta release

... in the order of their approximate order of priority IMO

* ... Merging all the small changes from master of course
* More documentation/examples/advanced use cases and linking this to the website
* Responsive option
* Doing more memory profiling and doing perf optimisation
* To mark the v2 change I would like to update and harmonise, refresh the general look and feel of the documentation + website + examples and provide more consistent branding. :ribbon:
* Adding visual regression testing (we can use media element backend and hard coded peak data to have this available even on travis/headless browser AFAIK) – but I haven't looked into this.
* Maybe switching to typescript – My IDE (visual studio code) does provide autocompletion and typechecking from the jsdoc comments, which I didn't anticipate before but which is pretty nice. – So this is a low-priority for me.

## Bugs which apply for master and next AFAIK

... I'm just including this here for a more complete roadmap

* Minimap plugin does not render complete waveform with large files https://github.com/katspaugh/wavesurfer.js/issues/994
* export image broken since MultiCanvas default https://github.com/katspaugh/wavesurfer.js/issues/1011
* Waveform display problem with timeline https://github.com/katspaugh/wavesurfer.js/issues/990
* Bad zooming performance when using timeline plugin https://github.com/katspaugh/wavesurfer.js/issues/967
* empty + redrawBuffer breaks player with pre-rendered peaks https://github.com/katspaugh/wavesurfer.js/issues/943
* partialRender with zoom = broken https://github.com/katspaugh/wavesurfer.js/issues/932

**Comment by mspae** [comment]
@entonbiba I think this can be closed – The limit will not be reached when https://github.com/katspaugh/wavesurfer.js/issues/1021 is merged.

**Pull request closed without merging**